sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -74,20 +74,98 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]] extends TreeNode[PlanT
-          case a: Attribute if a.semanticEquals(l) => r
+        val candidateConstraints = constraints - eq
+        inferredConstraints ++= candidateConstraints.map(_ transform {
+          case a: Attribute if a.semanticEquals(l) &&
+            !isRecursiveDeduction(r, aliasMap, equalExprSets) => r
         })
-        inferredConstraints ++= (constraints - eq).map(_ transform {
-          case a: Attribute if a.semanticEquals(r) => l
+        inferredConstraints ++= candidateConstraints.map(_ transform {
+          case a: Attribute if a.semanticEquals(r) &&
+            !isRecursiveDeduction(l, aliasMap, equalExprSets) => l
         })
       case _ => // No inference
     }
     inferredConstraints -- constraints
   }
 
+  /*
+   * Generate a sequence of expression sets from constraints, where each set stores an equivalence
+   * class of expressions. For example, Set(`a = b`, `b = c`, `e = f`) will generate the following
+   * expression sets: (Set(a, b, c), Set(e, f)). This will be used to search all expressions equal
+   * to an selected attribute.
+   */
+  private def generateEqualExpressionSets(
+      constraints: Set[Expression],
+      aliasMap: AttributeMap[Expression]): Seq[Set[Expression]] = {
+    var equalExprSets = Seq.empty[Set[Expression]]
+    constraints.foreach {
+      case eq @ EqualTo(l: Attribute, r: Attribute) =>
+        // Transform [[Alias]] to its child.
+        val left = aliasMap.getOrElse(l, l)
+        val right = aliasMap.getOrElse(r, r)
+        // Get the expression set for equivalence class of expressions.
+        val leftEqualSet = getEqualExprSet(left, equalExprSets).getOrElse(Set.empty[Expression])
+        val rightEqualSet = getEqualExprSet(right, equalExprSets).getOrElse(Set.empty[Expression])
+        if (!leftEqualSet.isEmpty && !rightEqualSet.isEmpty) {
+          // Combine the two sets.
+          equalExprSets = equalExprSets.diff(leftEqualSet :: rightEqualSet :: Nil) :+
+            (leftEqualSet ++ rightEqualSet)
+        } else if (!leftEqualSet.isEmpty) { // && rightEqualSet.isEmpty
+          // Update equivalence class of `left` expression.
+          equalExprSets = equalExprSets.diff(leftEqualSet :: Nil) :+ (leftEqualSet + right)
+        } else if (!rightEqualSet.isEmpty) { // && leftEqualSet.isEmpty
+          // Update equivalence class of `right` expression.
+          equalExprSets = equalExprSets.diff(rightEqualSet :: Nil) :+ (rightEqualSet + left)
+        } else { // leftEqualSet.isEmpty && rightEqualSet.isEmpty
+          // Create new equivalence class since both expression don't present in any classes.
+          equalExprSets = equalExprSets :+ Set(left, right)
+        }
+      case _ => // Skip
+    }
+
+    equalExprSets
+  }
+
+  /*
+   * Get all expressions equivalent to the selected expression.
+   */
+  private def getEqualExprSet(
+      expr: Expression,
+      equalExprSets: Seq[Set[Expression]]): Option[Set[Expression]] =
```

Comment:
we can just have this return `Set[Expression]` and fold the `.getOrElse(Set.empty[Expression])` logic inside